src/catalog/catalog.cpp

```diff
@@ -796,61 +796,140 @@ Catalog::~Catalog() {
 // FUNCTION
 //===--------------------------------------------------------------------===//
 
-void Catalog::AddFunction(
-    const std::string &name, const std::vector<type::TypeId> &argument_types,
-    const type::TypeId return_type,
-    type::Value (*func_ptr)(const std::vector<type::Value> &)) {
-  PL_ASSERT(functions_.count(name) == 0);
-  functions_[name] = FunctionData{name, argument_types, return_type, func_ptr};
+/*@brief   Add new built-in function
+ * 1. add the function infomation into pg_proc
+ * 2. register the function pointer in function::BuiltinFunction
+ * @param   name & argument_types   function name and arg types used in SQL
+ * @param   return_type   the return type
+ * @param   prolang       the oid of which language the function is
+ * @param   func_name     the function name in C++ source code (should be unique)
```

Comment:
Is the func_name value something that you allow the user to define? It looks like you need a unique name to identify the corresponding function_ptr ( to break ties when there are functions with the same name but different argument types).

Reply:
Yes, that's it. It is the user who defines the func_name.